core/storage/hamt/hamt.cpp

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/hamt/hamt.hpp"
+
+#include "codec/cbor/cbor.hpp"
+#include "common/which.hpp"
+#include "crypto/blake2/blake2b160.hpp"
+#include "crypto/murmur/murmur.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(fc::storage::hamt, HamtError, e) {
+  using fc::storage::hamt::HamtError;
+  switch (e) {
+    case HamtError::EXPECTED_CID:
+      return "Expected CID";
+    case HamtError::NOT_FOUND:
+      return "Not found";
+    case HamtError::MAX_DEPTH:
+      return "Max depth exceeded";
+    default:
+      return "Unknown error";
+  }
+}
+
+namespace fc::storage::hamt {
+  using fc::common::which;
+
+  const CID kDummyCid({}, {}, libp2p::multi::Multihash::create({}, {}).value());
+
+  // assuming 8-bit indices
+  auto keyToIndices(const std::string &key, int n = -1) {
+    std::vector<uint8_t> key_bytes;
```

Comment:
Code below seems to work. Ensure that argument of subspan is correct
```std::vector<uint8_t> key_bytes(key.begin(), key.end());
    auto hash = crypto::murmur::hash(key_bytes);
    auto span = gsl::make_span(hash).subspan(key.size() - n);
    return std::vector<size_t>(span.begin(), span.end());
```
Looks prettier.

Reply:
Actually it doesn't work, I checked. Something really weird happens here. 
In tests sometimes index goes out of bound. Without spans it can silently pass becoming a floating bug. Please recheck your implementation and tests and rewrite based on the code I specified.

Reply:
Thanks, didn't know about this posibility.